src/ClimaLSM.jl

```diff
@@ -1,329 +1,121 @@
-} <: AbstractModel{FT}
+    SM <: AbstractSoilModel{FT},
+    VM <: AbstractVegetationModel{FT},
+} <: AbstractLandModel{FT}
     configuration::CT
     soil::SM
     vegetation::VM
-    river::RM
-    snow::SNM
-    carbon::CM
 end
 
-function LandModel{FT}(;
+function RootSoilModel{FT}(;
     soil_args::NamedTuple = (;),
     vegetation_args::NamedTuple = (;),
-    river_args::NamedTuple = (;),
-    snow_args::NamedTuple = (;),
-    carbon_args::NamedTuple = (;),
 ) where {FT}
-    
-    configuration = RootSoilConfiguration{FT}() 
-    soil =
-        isempty(soil_args) ? NotIncluded{FT}(:soil) :
-        Soil.RichardsModel{FT}(; configuration = configuration, soil_args...)
-    vegetation =
-        isempty(vegetation_args) ? NotIncluded{FT}(:vegetation) :
-        Roots.RootsModel{FT}(;
-            configuration = configuration,
-            vegetation_args...,
-        )
-    river = NotIncluded{FT}(:river)
-    snow = NotIncluded{FT}(:snow)
-    carbon = NotIncluded{FT}(:carbon)
-    args = (configuration, soil, vegetation, river, snow, carbon)
-    return LandModel{FT, typeof.(args)...}(args...)
+    #somehow need to pass in soil version and veg version...
+    #configuration may depend on the versions of the soil and root models too!
+    configuration = Configurations.RootSoilConfiguration{FT}() 
+    soil = Soil.RichardsModel{FT}(; configuration = configuration, soil_args...)
+    vegetation = Roots.RootsModel{FT}(;
+                                      configuration = configuration,
+                                      vegetation_args...,
+                                      )
+    args = (configuration, soil, vegetation)
+    return RootSoilModel{FT, typeof.(args)...}(args...)
 end
 
 function initialize_interactions(
-    land::LandModel{FT, RootSoilConfiguration{FT}}) where {FT}
+    land::RootSoilModel{FT, RootSoilConfiguration{FT}}) where {FT}
     soil_coords = land.soil.coordinates
     return (root_extraction = similar(soil_coords),)
 end
 
-function initialize(land::LandModel)
-    Y_soil, p_soil, coords_soil = initialize(land.soil)
-    Y_vegetation, p_vegetation, coords_vegetation = initialize(land.vegetation)
-    Y_river, p_river, coords_river = initialize(land.river)
-    Y_snow, p_snow, coords_snow = initialize(land.snow)
-    Y_carbon, p_carbon, coords_carbon = initialize(land.carbon)
+
+land_components(land::RootSoilModel) = (:soil, :vegetation)
+
+function initialize(land::AbstractLandModel)
+    components = land_components(land)
+    coords_list = map(components) do (component)
+        Domains.coordinates(getproperty(land, component))
+    end
+    Y_state_list = map(zip(components,coords_list)) do (component, coords)
+        getproperty(initialize_prognostic(getproperty(land, component), coords), component)
+    end
+    p_state_list = map(zip(components,coords_list)) do (component, coords)
+        getproperty(initialize_auxiliary(getproperty(land, component), coords), component)
```

Comment:
I agree - these map/do blocks are hard to parse. I dont know exactly how mapping over a list of known size compares with a `for` loop over a list of known size. @charleskawczynski  had mentioned to me a few times that appending to a list in a `for` loop is not ideal because it is hard for the compiler (?) to allocate memory efficiently since it doesnt know how long the list will be, but `map` avoids that. Not sure, exactly, since in a `for` loop you also are looping over something of known size. We only need to do this list once, at initialization, so it doesnt matter much, but I was trying to practice that advice.


agreed that all the `get_property` are not nice. Let's revisit this entire `initialize` in a future PR, once we get some of this base functionality merged into main? I agree that for Y we can probably drop the ".component" in standalone mode, but still not sure about `p`.

Reply:
Looping is fine, I think my suggestion was to avoid `push!`ing, which can be expensive (especially for large generic arrays).

Reply:
Hm, I was thinking of this:
```
my_list = []
for x in [1,2,3,4]
push!(x, my_list)
end
```
vs
```
map(x-> x, [1,2,3,4])
```

Reply:
Yeah, IMO, the issue with `push!`ing is that it dynamically allocates on every `push!` call, whereas, IIUC, `map` will do a single allocation. If you know what the elements are:

```julia
my_list = Vector{Float64}(undef)
for i in (1,2,3,4)
    my_list[i] = ...
end
```
Then this also allocates once at `my_list = Vector{Float64}(undef)`.

Reply:
thanks!